Changes to bring the CSM SMD and OCHAMI SMD in sync in the go code

shunr-hpe opened the pull request:
These are the changes to make OCHAMI's go code work in in CSM. This includes test file changes to get the tests to pass in the CSM builds.

The CSM side of these changes are also being worked. In other words, the changes to bring OCHAMI into CSM.

alexlovelltroy commented:
I left detailed feedback in a few places, but I am overall quite pleased with how little difference there is.  A few areas for further discussion and testing needed.

alexlovelltroy commented:
There are also kafka errors in the goreleaser build that will need to be addressed

davidallendj commented:
> I left detailed feedback in a few places, but I am overall quite pleased with how little difference there is. A few areas for further discussion and testing needed.

Where are the comments? I don't see a review nor do a see any in this thread. Am I missing something?

davidallendj commented on cmd/smd/smd-api.go:
```diff
@@ -2551,21 +2551,23 @@ func (s *SmD) doRedfishEndpointsPost(w http.ResponseWriter, r *http.Request) {
-		if err != nil {
-			sendJsonError(w, http.StatusInternalServerError,
-				fmt.Sprintf("failed parsing post data (V2): %w", err))
+	if s.ochami {
```
This is a part of the parsing code that shouldn't affect any of the original CSM code. I'm guessing the intent here is to enable/disable all OpenCHAMI added features with the flag?

shunr-hpe replied:
Schema validation is a feature that I decided to deal with later. We may still enable it in CSM in the future.

My initial thoughts were that for a new project it is fine and good to add these additional checks. For an old project, like CSM, it is a little harder, the schema check could prevent something that someone is relying on as a "feature". However, I haven't looked into what this was doing.

I will consider this later (assuming this PR is approved without that part)

shunr-hpe commented on cmd/smd/msgbus.go:
```diff
@@ -0,0 +1,143 @@
+// MIT License
+//
+// (C) Copyright [2019-2021] Hewlett Packard Enterprise Development LP
```
I'm unsure which copyright can be used.
This file has been in CSM SMD for a long time and predates OpenCHAMI. My managers (HPE) and others will discuss what to do.
The original file is here:
https://github.com/Cray-HPE/hms-smd/blob/master/cmd/smd/msgbus.go

shunr-hpe commented on cmd/smd/rfevent-monitor.go:
```diff
@@ -0,0 +1,79 @@
+// MIT License
+//
+// (C) Copyright [2019-2021] Hewlett Packard Enterprise Development LP
```
I'm unsure which copyright can be used.
This file has been in CSM SMD for a long time and predates OpenCHAMI. My managers (HPE) and others will discuss what to do.
The original file is here:
https://github.com/Cray-HPE/hms-smd/blob/master/cmd/smd/rfevent-monitor.go

shunr-hpe commented on cmd/smd/routers.go:
```diff
@@ -59,12 +56,20 @@ func (s *SmD) NewRouter(publicRoutes []Route, protectedRoutes []Route) *chi.Mux
 		s.Logger(http.NotFoundHandler(), "NotFoundHandler")
 	}))
 
+	// Setup logger
```
I'll add a flag for zerolog and change ochami to openchami, 
I'll look into adding specific flags for the other uses of ochami and then remove ochami.

shunr-hpe commented on cmd/smd/smd-api_test.go:
```diff
@@ -870,7 +871,7 @@ func TestDoComponentGet(t *testing.T) {
 		errors.New("unexpected DB error"), // Could be any non HMSError
 		"x0c0s27",
-		json.RawMessage(`{"type":"about:blank","title":"Internal Server Error","detail":"failed to query DB.","status":500}
+		json.RawMessage(`{"type":"about:blank","title":"Internal Server Error","status":500}
```
This is a change to the unittest to get it to pass.
The change to the product code was originally made in the OpenCHAMI version of smd. 

OpenCHAMI made a change to the product code a while ago:
https://github.com/OpenCHAMI/smd/commit/da00d0313588510a49a97f31b923c2d63b17d00b

Current OpenCHAMI code:
https://github.com/OpenCHAMI/smd/blob/main/cmd/smd/response.go#L42-L44

Current CSM code:
https://github.com/Cray-HPE/hms-smd/blob/master/cmd/smd/response.go#L91-L94

I didn't mind that the "failed to query DB" message was removed from the error message. That message didn't seem to have much information. As such, I kept that product change and changed the unit test.

I'm willing to change the product code instead of the test, if people prefer the original error message.

alexlovelltroy replied:
Since these errors end up getting propagated through the call chain, I support the more detailed message as it provides a better reference for where to look when troubleshooting

shunr-hpe commented:
I have pushed a full new set of commits.
Here are the changes:
1. Signed all the commits (which is why a new set of commits was needed)
2. Added a build flag to include or exclude the message bus code. This fixes the goreleaser build failures.
3. Fixed the Makefile unittest target
4. Changed ochami to openchami
5. Added binaries to .gitignore
6. Ran go mod tidy
7. Added a module path check to determine defaults for the openchami and zerolog flags, and to test that the msg bus code didn't unexpectedly get included.

Here are some issues that I plan to consider and deal with later, unless anyone thinks that they most be dealt with in this PR.
1. Copyrights. HPE or OpenCHAMI.
2. Add other run time flags. We could possibly remove the openchami flag and add more specific flags.

davidallendj commented:
When running magellan with the csm-changes branch of SMD, I'm getting an error from SMD when doing a collect:
```
{"level":"error","error":"400: {\"type\":\"about:blank\",\"title\":\"Bad Request\",\"detail\":\"Invalid CompEthInterface IP Address\",\"status\":400}\n{\"type\":\"about:blank\",\"title\":\"Bad Request\",\"detail\":\"Invalid CompEthInterface IP Address\",\"status\":400}\n{\"type\":\"about:blank\",\"title\":\"Bad Request\",\"detail\":\"Invalid CompEthInterface IP Address\",\"status\":400}\n{\"type\":\"about:blank\",\"title\":\"Bad Request\",\"detail\":\"Invalid CompEthInterface IP Address\",\"status\":400}\n{\"type\":\"about:blank\",\"title\":\"Bad Request\",\"detail\":\"Invalid CompEthInterface IP Address\",\"status\":400}\n[{\"URI\":\"/hsm/v2/Inventory/RedfishEndpoints/x1000c1s7b0\"}]\n","time":"2025-04-07T12:44:34-06:00","message":"failed to add Redfish endpoint"}
```
I don't have this issue testing `magellan` with OpenCHAMI SMD v2.17.7.

shunr-hpe commented:
> When running magellan with the csm-changes branch of SMD, I'm getting an error from SMD when doing a collect:
> 
> ```
> {"level":"error","error":"400: {\"type\":\"about:blank\",\"title\":\"Bad Request\",\"detail\":\"Invalid CompEthInterface IP Address\",\"status\":400}\n{\"type\":\"about:blank\",\"title\":\"Bad Request\",\"detail\":\"Invalid CompEthInterface IP Address\",\"status\":400}\n{\"type\":\"about:blank\",\"title\":\"Bad Request\",\"detail\":\"Invalid CompEthInterface IP Address\",\"status\":400}\n{\"type\":\"about:blank\",\"title\":\"Bad Request\",\"detail\":\"Invalid CompEthInterface IP Address\",\"status\":400}\n{\"type\":\"about:blank\",\"title\":\"Bad Request\",\"detail\":\"Invalid CompEthInterface IP Address\",\"status\":400}\n[{\"URI\":\"/hsm/v2/Inventory/RedfishEndpoints/x1000c1s7b0\"}]\n","time":"2025-04-07T12:44:34-06:00","message":"failed to add Redfish endpoint"}
> ```
> 
> I don't have this issue testing `magellan` with OpenCHAMI SMD v2.17.7.

I get these same failures when using a commit that has already been submitted. In other words, when using a commit that isn't a part of this PR. If someone has time please confirm this, by testing with a build of SMD HEAD, or at least e7da43f8f9f9281f4a19a0f618e76d8af05731e9

In my testing both v2.17.7 and v2.18.0 work with megallan, however, the next PR after v2.18.0 does not work, or at least does not work when I test it. The PR is https://github.com/OpenCHAMI/smd/pull/49

davidallendj commented:
@shunr-hpe Thanks for pointing that out. I'll test using that commit and see what's causing the parsing to not work correctly tomorrow.

davidallendj commented:
I took a deeper dive and found that the issue is with https://github.com/OpenCHAMI/smd/commit/753031af6356233afff496e1e142ebecc35fdc6a which https://github.com/OpenCHAMI/smd/commit/e7da43f8f9f9281f4a19a0f618e76d8af05731e9 builds on. I'm working on fixing the issue and testing again with this branch.

davidallendj commented:
Can someone try applying this patch and see if it works?
```bash
diff --git a/cmd/smd/smd-api.go b/cmd/smd/smd-api.go
index 011b753..8b17a2c 100644
--- a/cmd/smd/smd-api.go
+++ b/cmd/smd/smd-api.go
@@ -2718,12 +2718,15 @@ func (s *SmD) parseRedfishEndpointDataV2(w http.ResponseWriter, data []byte, for
        var createCompEthInterfacesV2 = func(component base.Component, eths []schemas.EthernetInterface) {
                for _, eth := range eths {
                        // convert IP address from manager ethernet interface to IPAddressMapping
-                       ips := []sm.IPAddressMapping{sm.IPAddressMapping{IPAddr: eth.IP}}
-                       cei, err := sm.NewCompEthInterfaceV2(eth.Description, eth.MAC, component.ID, ips)
-                       if err != nil {
-                               sendJsonError(w, http.StatusBadRequest, err.Error())
-                               continue
-                       }
+                       var (
+                               ips = []sm.IPAddressMapping{sm.IPAddressMapping{IPAddr: eth.IP}}
+                               cei = &sm.CompEthInterfaceV2{
+                                       ID:      component.ID,
+                                       Desc:    eth.Description,
+                                       MACAddr: eth.MAC,
+                                       IPAddrs: ips,
+                               }
+                       )
                        err = s.db.InsertCompEthInterface(cei)
                        if err != nil {
                                if err == hmsds.ErrHMSDSDuplicateKey {
```

shunr-hpe commented:
> Can someone try applying this patch and see if it works?
> 
> ```shell
> diff --git a/cmd/smd/smd-api.go b/cmd/smd/smd-api.go
> index 011b753..d8f3e0d 100644
> --- a/cmd/smd/smd-api.go
> +++ b/cmd/smd/smd-api.go
> @@ -2718,10 +2718,18 @@ func (s *SmD) parseRedfishEndpointDataV2(w http.ResponseWriter, data []byte, for
>         var createCompEthInterfacesV2 = func(component base.Component, eths []schemas.EthernetInterface) {
>                 for _, eth := range eths {
>                         // convert IP address from manager ethernet interface to IPAddressMapping
> -                       ips := []sm.IPAddressMapping{sm.IPAddressMapping{IPAddr: eth.IP}}
> -                       cei, err := sm.NewCompEthInterfaceV2(eth.Description, eth.MAC, component.ID, ips)
> -                       if err != nil {
> -                               sendJsonError(w, http.StatusBadRequest, err.Error())
> +                       var (
> +                               ips = []sm.IPAddressMapping{sm.IPAddressMapping{IPAddr: eth.IP}}
> +                               cei = &sm.CompEthInterfaceV2{
> +                                       ID:      component.ID,
> +                                       Desc:    eth.Description,
> +                                       MACAddr: eth.MAC,
> +                                       IPAddrs: ips,
> +                               }
> +                       )
> +                       // check for required MAC address
> +                       if cei.MACAddr == "" {
> +                               sendJsonError(w, http.StatusBadRequest, sm.ErrCompEthInterfaceBadMAC.Error())
>                                 continue
>                         }
>                         err = s.db.InsertCompEthInterface(cei)
> ```
Here are these ^^^ changes:
https://github.com/OpenCHAMI/smd/compare/csm-changes...csm-changes-patched-for-parse-redfish-endpoint

My first pass at testing produces a different error
```
/tmp/nobody/magellan $ magellan collect --host http://smd:27779 --access-token "$ACCESS_TOKEN"
...
{"level":"error","error":"409: {\"type\":\"about:blank\",\"title\":\"Conflict\",\"detail\":\"operation would conflict with an existing component ethernet interface that has the same MAC address.\",\"status\":409}\n{\"type\":\"about:blank\",\"title\":\"Conflict\",\"detail\":\"operation would conflict with an existing component ethernet interface that has the same MAC address.\",\"status\":409}\n{\"type\":\"about:blank\",\"title\":\"Conflict\",\"detail\":\"operation would conflict with an existing component ethernet interface that has the same MAC address.\",\"status\":409}\n[{\"URI\":\"/hsm/v2/Inventory/RedfishEndpoints/x1000c1s7b2\"}]\n","time":"2025-04-11T18:39:49Z","message":"failed to add Redfish endpoint"}
```

synackd commented:
> > Can someone try applying this patch and see if it works?
> > ```shell
> > diff --git a/cmd/smd/smd-api.go b/cmd/smd/smd-api.go
> > index 011b753..d8f3e0d 100644
> > --- a/cmd/smd/smd-api.go
> > +++ b/cmd/smd/smd-api.go
> > @@ -2718,10 +2718,18 @@ func (s *SmD) parseRedfishEndpointDataV2(w http.ResponseWriter, data []byte, for
> >         var createCompEthInterfacesV2 = func(component base.Component, eths []schemas.EthernetInterface) {
> >                 for _, eth := range eths {
> >                         // convert IP address from manager ethernet interface to IPAddressMapping
> > -                       ips := []sm.IPAddressMapping{sm.IPAddressMapping{IPAddr: eth.IP}}
> > -                       cei, err := sm.NewCompEthInterfaceV2(eth.Description, eth.MAC, component.ID, ips)
> > -                       if err != nil {
> > -                               sendJsonError(w, http.StatusBadRequest, err.Error())
> > +                       var (
> > +                               ips = []sm.IPAddressMapping{sm.IPAddressMapping{IPAddr: eth.IP}}
> > +                               cei = &sm.CompEthInterfaceV2{
> > +                                       ID:      component.ID,
> > +                                       Desc:    eth.Description,
> > +                                       MACAddr: eth.MAC,
> > +                                       IPAddrs: ips,
> > +                               }
> > +                       )
> > +                       // check for required MAC address
> > +                       if cei.MACAddr == "" {
> > +                               sendJsonError(w, http.StatusBadRequest, sm.ErrCompEthInterfaceBadMAC.Error())
> >                                 continue
> >                         }
> >                         err = s.db.InsertCompEthInterface(cei)
> > ```
> 
> Here are these ^^^ changes: [csm-changes...csm-changes-patched-for-parse-redfish-endpoint](https://github.com/OpenCHAMI/smd/compare/csm-changes...csm-changes-patched-for-parse-redfish-endpoint)

While checking the MAC address, the format of the address should also be checked as in:

https://github.com/OpenCHAMI/smd/blob/main/pkg/sm/compethinterfaces.go#L137-L141

> My first pass at testing produces a different error
> 
> ```
> /tmp/nobody/magellan $ magellan collect --host http://smd:27779 --access-token "$ACCESS_TOKEN"
> ...
> {"level":"error","error":"409: {\"type\":\"about:blank\",\"title\":\"Conflict\",\"detail\":\"operation would conflict with an existing component ethernet interface that has the same MAC address.\",\"status\":409}\n{\"type\":\"about:blank\",\"title\":\"Conflict\",\"detail\":\"operation would conflict with an existing component ethernet interface that has the same MAC address.\",\"status\":409}\n{\"type\":\"about:blank\",\"title\":\"Conflict\",\"detail\":\"operation would conflict with an existing component ethernet interface that has the same MAC address.\",\"status\":409}\n[{\"URI\":\"/hsm/v2/Inventory/RedfishEndpoints/x1000c1s7b2\"}]\n","time":"2025-04-11T18:39:49Z","message":"failed to add Redfish endpoint"}
> ```

Are you starting clean?

shunr-hpe commented:
Yes, for my testing it was clean / all new docker containers and volumes.
My testing was using csm-rie: https://github.com/OpenCHAMI/csm-redfish-interface-emulator
```
docker compose -f base.yml -f postgres.yml -f jwt-security.yml -f haproxy-api-gateway.yml -f  openchami-svcs.yml -f autocert.yml -f coredhcp.yml -f configurator.yml -f computes.yml up -d
```
where computes.yml contains
```
services:
  rf-x0c0s1b0:
    container_name: rf-x0c0s1b0
    hostname: x0c0s1b0
    image: ghcr.io/openchami/csm-rie:latest
    environment:
      - MOCKUPFOLDER=EX235a
      - MAC_SCHEMA=Mountain
      - XNAME=x0c0s1b0
      - PORT=443
    networks:
      internal:
        aliases:
          - x0c0s1b0
  rf-x0c0s2b0:
    container_name: rf-x0c0s2b0
    hostname: x0c0s2b0
    image: ghcr.io/openchami/csm-rie:latest
    environment:
      - MOCKUPFOLDER=EX235a
      - MAC_SCHEMA=Mountain
      - XNAME=x0c0s2b0
      - PORT=443
    networks:
      internal:
        aliases:
          - x0c0s2b0
  rf-x0c0s3b0:
    container_name: rf-x0c0s3b0
    hostname: x0c0s3b0
    image: ghcr.io/openchami/csm-rie:latest
    environment:
      - MOCKUPFOLDER=EX235a
      - MAC_SCHEMA=Mountain
      - XNAME=x0c0s3b0
      - PORT=443
    networks:
      internal:
        aliases:
          - x0c0s3b0
  rf-x0c0s4b0:
    container_name: rf-x0c0s4b0
    hostname: x0c0s4b0
    image: ghcr.io/openchami/csm-rie:latest
    environment:
      - MOCKUPFOLDER=EX235a
      - MAC_SCHEMA=Mountain
      - XNAME=x0c0s4b0
      - PORT=443
    networks:
      internal:
        aliases:
          - x0c0s4b0
```

davidallendj commented:
> While checking the MAC address, the format of the address should also be checked as in:
>
> https://github.com/OpenCHAMI/smd/blob/main/pkg/sm/compethinterfaces.go#L137-L141

Should the ID be the MAC address? Right now, it's being set to the component ID.

Here's where the `CompEthInterface` ID is being set initially:
https://github.com/OpenCHAMI/smd/blob/main/cmd/smd/smd-api.go#L2722 

Here's where it's being set to the MAC address.
https://github.com/OpenCHAMI/smd/blob/main/pkg/sm/compethinterfaces.go#L138

> My first pass at testing produces a different error
> 
> /tmp/nobody/magellan $ magellan collect --host http://smd:27779 --access-token "$ACCESS_TOKEN"
>  ...
>  {"level":"error","error":"409: {\"type\":\"about:blank\",\"title\":\"Conflict\",\"detail\":\"operation would conflict with an existing component ethernet interface that has the same MAC address.\",\"status\":409}\n{\"type\":\"about:blank\",\"title\":\"Conflict\",\"detail\":\"operation would conflict with an existing component ethernet interface that has the same MAC address.\",\"status\":409}\n{\"type\":\"about:blank\",\"title\":\"Conflict\",\"detail\":\"operation would conflict with an existing component ethernet interface that has the same MAC address.\",\"status\":409}\n[{\"URI\":\"/hsm/v2/Inventory/RedfishEndpoints/x1000c1s7b2\"}]\n","time":"2025-04-11T18:39:49Z","message":"failed to add Redfish endpoint"}

Looks like that's failing here: https://github.com/OpenCHAMI/smd/blob/main/cmd/smd/smd-api.go#L2727
Error output further down here: https://github.com/OpenCHAMI/smd/blob/main/cmd/smd/smd-api.go#L2754

@shunr-hpe Can you do `/tmp/nobody/magellan $ magellan collect --host http://smd:27779 --access-token "$ACCESS_TOKEN" -v` and post the Magellan output?

shunr-hpe commented:
@davidallendj 
```
/tmp/nobody/magellan $ magellan collect --host http://smd:27779 --access-token "$ACCESS_TOKEN" -v
```
```
{"level":"info","time":"2025-04-11T19:41:13Z","message":"found ServiceRoot RootService. Redfish Version 1.2.0"}
{"level":"info","time":"2025-04-11T19:41:13Z","message":"found ServiceRoot RootService. Redfish Version 1.2.0"}
{"level":"info","time":"2025-04-11T19:41:13Z","message":"found ServiceRoot RootService. Redfish Version 1.2.0"}
{"level":"info","time":"2025-04-11T19:41:13Z","message":"found ServiceRoot RootService. Redfish Version 1.2.0"}
{"level":"info","time":"2025-04-11T19:41:14Z","message":"found 4 chassis in ServiceRoot"}
{"level":"info","time":"2025-04-11T19:41:14Z","message":"found 4 chassis in ServiceRoot"}
{"level":"info","time":"2025-04-11T19:41:14Z","message":"found 0 systems in chassis Mezz0"}
{"level":"info","time":"2025-04-11T19:41:14Z","message":"found 0 systems in chassis Mezz1"}
{"level":"info","time":"2025-04-11T19:41:14Z","message":"found 0 systems in chassis Enclosure"}
{"level":"info","time":"2025-04-11T19:41:14Z","message":"found 0 systems in chassis Node0"}
{"level":"info","time":"2025-04-11T19:41:14Z","message":"found 4 chassis in ServiceRoot"}
{"level":"info","time":"2025-04-11T19:41:14Z","message":"found 0 systems in chassis Mezz0"}
{"level":"info","time":"2025-04-11T19:41:14Z","message":"found 0 systems in chassis Mezz1"}
{"level":"info","time":"2025-04-11T19:41:14Z","message":"found 0 systems in chassis Enclosure"}
{"level":"info","time":"2025-04-11T19:41:14Z","message":"found 0 systems in chassis Node0"}
{"level":"info","time":"2025-04-11T19:41:14Z","message":"found 4 chassis in ServiceRoot"}
{"level":"info","time":"2025-04-11T19:41:14Z","message":"found 0 systems in chassis Mezz0"}
{"level":"info","time":"2025-04-11T19:41:14Z","message":"found 0 systems in chassis Enclosure"}
{"level":"info","time":"2025-04-11T19:41:14Z","message":"found 0 systems in chassis Mezz1"}
{"level":"info","time":"2025-04-11T19:41:14Z","message":"found 0 systems in chassis Node0"}
{"level":"info","time":"2025-04-11T19:41:14Z","message":"found 0 systems in chassis Mezz0"}
{"level":"info","time":"2025-04-11T19:41:14Z","message":"found 0 systems in chassis Mezz1"}
{"level":"info","time":"2025-04-11T19:41:14Z","message":"found 0 systems in chassis Enclosure"}
{"level":"info","time":"2025-04-11T19:41:14Z","message":"found 0 systems in chassis Node0"}
{"level":"info","time":"2025-04-11T19:41:14Z","message":"found 1 systems in ServiceRoot"}
{"level":"info","time":"2025-04-11T19:41:14Z","message":"found 1 systems in ServiceRoot"}
{"level":"info","time":"2025-04-11T19:41:14Z","message":"found 1 systems in ServiceRoot"}
{"level":"info","time":"2025-04-11T19:41:14Z","message":"found 1 systems in ServiceRoot"}
{"level":"info","time":"2025-04-11T19:41:14Z","message":"found ServiceRoot RootService. Redfish Version 1.2.0"}
{"level":"info","time":"2025-04-11T19:41:14Z","message":"found ServiceRoot RootService. Redfish Version 1.2.0"}
{"level":"info","time":"2025-04-11T19:41:14Z","message":"found ServiceRoot RootService. Redfish Version 1.2.0"}
{"level":"info","time":"2025-04-11T19:41:14Z","message":"found ServiceRoot RootService. Redfish Version 1.2.0"}
{"level":"warn","error":"no ethernet interfaces found with IP address","time":"2025-04-11T19:41:14Z","message":"failed to find MAC address with IP 'https://172.18.0.3'"}
{
    "FQDN": "https://172.18.0.3",
    "ID": "x1000c1s7b1",
    "MACRequired": true,
    "Managers": [
        {
            "uri": "https://172.18.0.3:443/redfish/v1/Managers/BMC",
            "name": "BMC",
            "description": "Shasta Manager",
            "type": "EnclosureManager"
        },
        {
            "uri": "https://172.18.0.3:443/redfish/v1/Managers/MCU0",
            "name": "MCU0",
            "description": "Microcontroller Unit for Mezz0",
            "type": "AuxiliaryController"
        },
        {
            "uri": "https://172.18.0.3:443/redfish/v1/Managers/MCU1",
            "name": "MCU1",
            "description": "Microcontroller Unit for Mezz1",
            "type": "AuxiliaryController"
        }
    ],
    "Name": "",
    "RediscoverOnUpdate": false,
    "SchemaVersion": 1,
    "Systems": [
        {
            "uri": "https://172.18.0.3:443/redfish/v1/Systems/Node0",
            "manufacturer": "HPE",
            "system_type": "Physical",
            "name": "Node0",
            "model": "HPE CRAY EX235a",
            "serial": "GOR4904808651",
            "bios_version": "ex235a.bios-1.3.6",
            "ethernet_interfaces": [
                {
                    "uri": "https://172.18.0.3:443/redfish/v1/Systems/Node0/EthernetInterfaces/HPCNet2",
                    "mac": "Not Available",
                    "description": "SS11 200Gb 2P NIC Mezz REV02 (HSN)"
                },
                {
                    "uri": "https://172.18.0.3:443/redfish/v1/Systems/Node0/EthernetInterfaces/ManagementEthernet",
                    "mac": "02:00:00:00:33:00",
                    "description": "Node Maintenance Network"
                },
                {
                    "uri": "https://172.18.0.3:443/redfish/v1/Systems/Node0/EthernetInterfaces/HPCNet1",
                    "mac": "Not Available",
                    "description": "SS11 200Gb 2P NIC Mezz REV02 (HSN)"
                },
                {
                    "uri": "https://172.18.0.3:443/redfish/v1/Systems/Node0/EthernetInterfaces/HPCNet3",
                    "mac": "Not Available",
                    "description": "SS11 200Gb 2P NIC Mezz REV02 (HSN)"
                },
                {
                    "uri": "https://172.18.0.3:443/redfish/v1/Systems/Node0/EthernetInterfaces/HPCNet0",
                    "mac": "Not Available",
                    "description": "SS11 200Gb 2P NIC Mezz REV02 (HSN)"
                }
            ],
            "power_state": "On",
            "processor_count": 9,
            "processor_type": "AMD INSTINCT MI200 (MCM) OAM LC",
            "memory_total": 512
        }
    ],
    "Type": "",
    "User": ""
}
{"level":"warn","error":"no ethernet interfaces found with IP address","time":"2025-04-11T19:41:14Z","message":"failed to find MAC address with IP 'https://172.18.0.4'"}
{
    "FQDN": "https://172.18.0.4",
    "ID": "x1000c1s7b2",
    "MACRequired": true,
    "Managers": [
        {
            "uri": "https://172.18.0.4:443/redfish/v1/Managers/BMC",
            "name": "BMC",
            "description": "Shasta Manager",
            "type": "EnclosureManager"
        },
        {
            "uri": "https://172.18.0.4:443/redfish/v1/Managers/MCU0",
            "name": "MCU0",
            "description": "Microcontroller Unit for Mezz0",
            "type": "AuxiliaryController"
        },
        {
            "uri": "https://172.18.0.4:443/redfish/v1/Managers/MCU1",
            "name": "MCU1",
            "description": "Microcontroller Unit for Mezz1",
            "type": "AuxiliaryController"
        }
    ],
    "Name": "",
    "RediscoverOnUpdate": false,
    "SchemaVersion": 1,
    "Systems": [
        {
            "uri": "https://172.18.0.4:443/redfish/v1/Systems/Node0",
            "manufacturer": "HPE",
            "system_type": "Physical",
            "name": "Node0",
            "model": "HPE CRAY EX235a",
            "serial": "PPB1843794187",
            "bios_version": "ex235a.bios-1.3.6",
            "ethernet_interfaces": [
                {
                    "uri": "https://172.18.0.4:443/redfish/v1/Systems/Node0/EthernetInterfaces/HPCNet2",
                    "mac": "Not Available",
                    "description": "SS11 200Gb 2P NIC Mezz REV02 (HSN)"
                },
                {
                    "uri": "https://172.18.0.4:443/redfish/v1/Systems/Node0/EthernetInterfaces/HPCNet1",
                    "mac": "Not Available",
                    "description": "SS11 200Gb 2P NIC Mezz REV02 (HSN)"
                },
                {
                    "uri": "https://172.18.0.4:443/redfish/v1/Systems/Node0/EthernetInterfaces/ManagementEthernet",
                    "mac": "02:00:00:00:32:00",
                    "description": "Node Maintenance Network"
                },
                {
                    "uri": "https://172.18.0.4:443/redfish/v1/Systems/Node0/EthernetInterfaces/HPCNet3",
                    "mac": "Not Available",
                    "description": "SS11 200Gb 2P NIC Mezz REV02 (HSN)"
                },
                {
                    "uri": "https://172.18.0.4:443/redfish/v1/Systems/Node0/EthernetInterfaces/HPCNet0",
                    "mac": "Not Available",
                    "description": "SS11 200Gb 2P NIC Mezz REV02 (HSN)"
                }
            ],
            "power_state": "On",
            "processor_count": 9,
            "processor_type": "AMD INSTINCT MI200 (MCM) OAM LC",
            "memory_total": 512
        }
    ],
    "Type": "",
    "User": ""
}
{"level":"warn","error":"no ethernet interfaces found with IP address","time":"2025-04-11T19:41:14Z","message":"failed to find MAC address with IP 'https://172.18.0.5'"}
{
    "FQDN": "https://172.18.0.5",
    "ID": "x1000c1s7b3",
    "MACRequired": true,
    "Managers": [
        {
            "uri": "https://172.18.0.5:443/redfish/v1/Managers/BMC",
            "name": "BMC",
            "description": "Shasta Manager",
            "type": "EnclosureManager"
        },
        {
            "uri": "https://172.18.0.5:443/redfish/v1/Managers/MCU0",
            "name": "MCU0",
            "description": "Microcontroller Unit for Mezz0",
            "type": "AuxiliaryController"
        },
        {
            "uri": "https://172.18.0.5:443/redfish/v1/Managers/MCU1",
            "name": "MCU1",
            "description": "Microcontroller Unit for Mezz1",
            "type": "AuxiliaryController"
        }
    ],
    "Name": "",
    "RediscoverOnUpdate": false,
    "SchemaVersion": 1,
    "Systems": [
        {
            "uri": "https://172.18.0.5:443/redfish/v1/Systems/Node0",
            "manufacturer": "HPE",
            "system_type": "Physical",
            "name": "Node0",
            "model": "HPE CRAY EX235a",
            "serial": "VZN5570323444",
            "bios_version": "ex235a.bios-1.3.6",
            "ethernet_interfaces": [
                {
                    "uri": "https://172.18.0.5:443/redfish/v1/Systems/Node0/EthernetInterfaces/HPCNet1",
                    "mac": "Not Available",
                    "description": "SS11 200Gb 2P NIC Mezz REV02 (HSN)"
                },
                {
                    "uri": "https://172.18.0.5:443/redfish/v1/Systems/Node0/EthernetInterfaces/HPCNet2",
                    "mac": "Not Available",
                    "description": "SS11 200Gb 2P NIC Mezz REV02 (HSN)"
                },
                {
                    "uri": "https://172.18.0.5:443/redfish/v1/Systems/Node0/EthernetInterfaces/ManagementEthernet",
                    "mac": "02:00:00:00:31:00",
                    "description": "Node Maintenance Network"
                },
                {
                    "uri": "https://172.18.0.5:443/redfish/v1/Systems/Node0/EthernetInterfaces/HPCNet3",
                    "mac": "Not Available",
                    "description": "SS11 200Gb 2P NIC Mezz REV02 (HSN)"
                },
                {
                    "uri": "https://172.18.0.5:443/redfish/v1/Systems/Node0/EthernetInterfaces/HPCNet0",
                    "mac": "Not Available",
                    "description": "SS11 200Gb 2P NIC Mezz REV02 (HSN)"
                }
            ],
            "power_state": "On",
            "processor_count": 9,
            "processor_type": "AMD INSTINCT MI200 (MCM) OAM LC",
            "memory_total": 512
        }
    ],
    "Type": "",
    "User": ""
}
{"level":"warn","error":"no ethernet interfaces found with IP address","time":"2025-04-11T19:41:14Z","message":"failed to find MAC address with IP 'https://172.18.0.2'"}
{
    "FQDN": "https://172.18.0.2",
    "ID": "x1000c1s7b0",
    "MACRequired": true,
    "Managers": [
        {
            "uri": "https://172.18.0.2:443/redfish/v1/Managers/BMC",
            "name": "BMC",
            "description": "Shasta Manager",
            "type": "EnclosureManager"
        },
        {
            "uri": "https://172.18.0.2:443/redfish/v1/Managers/MCU0",
            "name": "MCU0",
            "description": "Microcontroller Unit for Mezz0",
            "type": "AuxiliaryController"
        },
        {
            "uri": "https://172.18.0.2:443/redfish/v1/Managers/MCU1",
            "name": "MCU1",
            "description": "Microcontroller Unit for Mezz1",
            "type": "AuxiliaryController"
        }
    ],
    "Name": "",
    "RediscoverOnUpdate": false,
    "SchemaVersion": 1,
    "Systems": [
        {
            "uri": "https://172.18.0.2:443/redfish/v1/Systems/Node0",
            "manufacturer": "HPE",
            "system_type": "Physical",
            "name": "Node0",
            "model": "HPE CRAY EX235a",
            "serial": "MCS5714683928",
            "bios_version": "ex235a.bios-1.3.6",
            "ethernet_interfaces": [
                {
                    "uri": "https://172.18.0.2:443/redfish/v1/Systems/Node0/EthernetInterfaces/HPCNet2",
                    "mac": "Not Available",
                    "description": "SS11 200Gb 2P NIC Mezz REV02 (HSN)"
                },
                {
                    "uri": "https://172.18.0.2:443/redfish/v1/Systems/Node0/EthernetInterfaces/HPCNet1",
                    "mac": "Not Available",
                    "description": "SS11 200Gb 2P NIC Mezz REV02 (HSN)"
                },
                {
                    "uri": "https://172.18.0.2:443/redfish/v1/Systems/Node0/EthernetInterfaces/ManagementEthernet",
                    "mac": "02:00:00:00:34:00",
                    "description": "Node Maintenance Network"
                },
                {
                    "uri": "https://172.18.0.2:443/redfish/v1/Systems/Node0/EthernetInterfaces/HPCNet3",
                    "mac": "Not Available",
                    "description": "SS11 200Gb 2P NIC Mezz REV02 (HSN)"
                },
                {
                    "uri": "https://172.18.0.2:443/redfish/v1/Systems/Node0/EthernetInterfaces/HPCNet0",
                    "mac": "Not Available",
                    "description": "SS11 200Gb 2P NIC Mezz REV02 (HSN)"
                }
            ],
            "power_state": "On",
            "processor_count": 9,
            "processor_type": "AMD INSTINCT MI200 (MCM) OAM LC",
            "memory_total": 512
        }
    ],
    "Type": "",
    "User": ""
}
{"level":"error","error":"409: {\"type\":\"about:blank\",\"title\":\"Conflict\",\"detail\":\"operation would conflict with an existing component ethernet interface that has the same MAC address.\",\"status\":409}\n{\"type\":\"about:blank\",\"title\":\"Conflict\",\"detail\":\"operation would conflict with an existing component ethernet interface that has the same MAC address.\",\"status\":409}\n{\"type\":\"about:blank\",\"title\":\"Conflict\",\"detail\":\"operation would conflict with an existing component ethernet interface that has the same MAC address.\",\"status\":409}\n{\"type\":\"about:blank\",\"title\":\"Conflict\",\"detail\":\"operation would conflict with an existing component ethernet interface that has the same MAC address.\",\"status\":409}\n[{\"URI\":\"/hsm/v2/Inventory/RedfishEndpoints/x1000c1s7b1\"}]\n","time":"2025-04-11T19:41:14Z","message":"failed to add Redfish endpoint"}
{"level":"error","error":"409: {\"type\":\"about:blank\",\"title\":\"Conflict\",\"detail\":\"operation would conflict with an existing component ethernet interface that has the same MAC address.\",\"status\":409}\n{\"type\":\"about:blank\",\"title\":\"Conflict\",\"detail\":\"operation would conflict with an existing component ethernet interface that has the same MAC address.\",\"status\":409}\n{\"type\":\"about:blank\",\"title\":\"Conflict\",\"detail\":\"operation would conflict with an existing component ethernet interface that has the same MAC address.\",\"status\":409}\n{\"type\":\"about:blank\",\"title\":\"Conflict\",\"detail\":\"operation would conflict with an existing component ethernet interface that has the same MAC address.\",\"status\":409}\n[{\"URI\":\"/hsm/v2/Inventory/RedfishEndpoints/x1000c1s7b3\"}]\n","time":"2025-04-11T19:41:14Z","message":"failed to add Redfish endpoint"}
{"level":"error","error":"409: {\"type\":\"about:blank\",\"title\":\"Conflict\",\"detail\":\"operation would conflict with an existing component ethernet interface that has the same MAC address.\",\"status\":409}\n{\"type\":\"about:blank\",\"title\":\"Conflict\",\"detail\":\"operation would conflict with an existing component ethernet interface that has the same MAC address.\",\"status\":409}\n{\"type\":\"about:blank\",\"title\":\"Conflict\",\"detail\":\"operation would conflict with an existing component ethernet interface that has the same MAC address.\",\"status\":409}\n{\"type\":\"about:blank\",\"title\":\"Conflict\",\"detail\":\"operation would conflict with an existing component ethernet interface that has the same MAC address.\",\"status\":409}\n[{\"URI\":\"/hsm/v2/Inventory/RedfishEndpoints/x1000c1s7b0\"}]\n","time":"2025-04-11T19:41:14Z","message":"failed to add Redfish endpoint"}
{"level":"error","error":"409: {\"type\":\"about:blank\",\"title\":\"Conflict\",\"detail\":\"operation would conflict with an existing component ethernet interface that has the same MAC address.\",\"status\":409}\n{\"type\":\"about:blank\",\"title\":\"Conflict\",\"detail\":\"operation would conflict with an existing component ethernet interface that has the same MAC address.\",\"status\":409}\n{\"type\":\"about:blank\",\"title\":\"Conflict\",\"detail\":\"operation would conflict with an existing component ethernet interface that has the same MAC address.\",\"status\":409}\n[{\"URI\":\"/hsm/v2/Inventory/RedfishEndpoints/x1000c1s7b2\"}]\n","time":"2025-04-11T19:41:14Z","message":"failed to add Redfish endpoint"}
```

davidallendj commented:
I'm wondering if the "Not Available" MAC addresses are causing the conflict somehow. I don't see any checks in `NewCompEthInterfaceV2` for it and I didn't see a "Not Available" string being checked anywhere. My guess is that's what is causing the "same MAC address" error.

synackd commented:
> > While checking the MAC address, the format of the address should also be checked as in:
> > https://github.com/OpenCHAMI/smd/blob/main/pkg/sm/compethinterfaces.go#L137-L141
> 
> Should the ID be the MAC address? Right now, it's being set to the component ID.

The ID is the MAC address without colons.

> Here's where the `CompEthInterface` ID is being set initially: https://github.com/OpenCHAMI/smd/blob/main/cmd/smd/smd-api.go#L2722
> 
> Here's where it's being set to the MAC address. https://github.com/OpenCHAMI/smd/blob/main/pkg/sm/compethinterfaces.go#L138

Yes, the link I sent is the spot in that function that verifies the MAC address and sets the ID. If we don't call this function, I'm suggesting that the lines I linked be copied into where the function call was in order to keep the verification/ID setting.

davidallendj commented:
I think I'm going to open a separate PR to address the Magellan-related issue to keep from blocking this PR.

davidallendj commented on cmd/smd/main_test.go:
```diff
@@ -0,0 +1,53 @@
+)
+
+func TestSmdFlavor(t *testing.T) {
+	name := "SmdFlavor"
```
Is this declaration necessary? I don't see it being set anywhere and it looks like it is only used in the error messages within this function's scope.

shunr-hpe replied:
Thanks, I've removed the variable.

shunr-hpe commented on cmd/smd/smd-api.go:
```diff
@@ -2260,14 +2260,14 @@ func (s *SmD) doRedfishEndpointPut(w http.ResponseWriter, r *http.Request) {
 		if err != nil {
 			sendJsonError(w, http.StatusInternalServerError,
-				fmt.Sprintf("failed parsing post data: %w", err))
+				fmt.Sprintf("failed parsing post data: %v", err))
```
I agree. I've fixed some of the cases in smd-api.go. There are some other places that need more work.

The http headers can only be written once.

golang's http2 server.go WriteHeader function documentation says:
```
	// The provided code must be a valid HTTP 1xx-5xx status code.
	// Any number of 1xx headers may be written, followed by at most
	// one 2xx-5xx header. 1xx headers are sent immediately, but 2xx-5xx
	// headers may be buffered. Use the Flusher interface to send
	// buffered data. The header map is cleared when 2xx-5xx headers are
	// sent, but not with 1xx headers.
```

shunr-hpe commented on cmd/smd/routers.go:
```diff
@@ -48,23 +48,25 @@ type Route struct {
 	router.Use(middleware.Recoverer)
 	router.Use(middleware.StripSlashes)
-	router.Use(openchami_logger.OpenCHAMILogger(logger))
+	if s.zerolog {
```
The rationale for having a separate config variable is so that the features can be enabled incrementally in CSM in the future. This way the CSM version of smd can eventually get closer to the OpenCHAMI version.

davidallendj reviewed (approved):
LGTM
